File: samples/bot-auth0-adaptivecard/nodejs/controllers/AuthController.js
const express = require('express');
const jwt = require('jsonwebtoken');
const TokenStore = require('../services/TokenStore');
const fetch = require('node-fetch');

const router = express.Router();

router.get('/callback', async (req, res) => {
    const { code, state } = req.query;

    if (!code || !state) {
        return res.status(400).send('Missing code or state in the callback');
    }

    try {
        const response = await fetch(`https://${process.env.AUTH0_DOMAIN}/oauth/token`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                grant_type: 'authorization_code',
                client_id: process.env.AUTH0_CLIENT_ID,
                client_secret: process.env.AUTH0_CLIENT_SECRET,
                code,
                redirect_uri: `${process.env.APP_URL}/api/auth/callback`
            })
        });

        if (!response.ok) {
            throw new Error('Failed to exchange code for token');
        }

        const data = await response.json();
        const { access_token } = data;

        if (access_token) {
            TokenStore.setToken(state, access_token);
            res.redirect('/public/auth-end.html');
        } else {
            res.status(400).send('No access token received');
        }
    } catch (err) {
        console.error('Auth callback error:', err);
        res.status(500).send('Authentication failed');
    }
});

module.exports = router;

File: samples/bot-auth0-adaptivecard/csharp/Controllers/AuthController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Configuration;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using System.Web;

namespace Microsoft.BotBuilderSamples.Controllers
{
    [Route("api/auth")]
    [ApiController]
    public class AuthController : ControllerBase
    {
        private readonly IConfiguration _config;
        private readonly IHttpClientFactory _clientFactory;

        public AuthController(IConfiguration config, IHttpClientFactory clientFactory)
        {
            _config = config;
            _clientFactory = clientFactory;
        }

        [HttpGet("callback")]
        public async Task<IActionResult> Callback(
        [FromQuery] string code,
        [FromQuery] string state,
        [FromServices] TokenStore tokenStore)
        {
            var client = _clientFactory.CreateClient();
            var tokenEndpoint = $"https://{_config["Auth0:Domain"]}/oauth/token";

            var response = await client.PostAsync(tokenEndpoint, new FormUrlEncodedContent(new Dictionary<string, string>
                            {
                                {"grant_type", "authorization_code"},
                                {"client_id", _config["Auth0:ClientId"]},
                                {"client_secret", _config["Auth0:ClientSecret"]},
                                {"code", code},
                                {"redirect_uri", _config["ApplicationUrl"] + "/api/auth/callback"}
                            }));

            if (!response.IsSuccessStatusCode)
            {
                return BadRequest("Token exchange failed");
            }

            var content = await response.Content.ReadAsStringAsync();
            var tokenResult = JsonSerializer.Deserialize<Auth0TokenResponse>(content);

            tokenStore.SaveToken(state, tokenResult.access_token);

            return Redirect($"/auth-end.html?token={HttpUtility.UrlEncode(content)}");
        }


        public class TokenStore
        {
            private readonly Dictionary<string, string> _tokens = new();

            public void SaveToken(string userId, string token)
            {
                _tokens[userId] = token;
            }

            public bool TryGetToken(string userId, out string token)
            {
                return _tokens.TryGetValue(userId, out token);
            }

            public void RemoveToken(string userId)
            {
                _tokens.Remove(userId);
            }
        }

        public class Auth0TokenResponse
        {
            public string access_token { get; set; }
            public string id_token { get; set; }
            public string scope { get; set; }
            public int expires_in { get; set; }
            public string token_type { get; set; }
        }
    }
}
